### ISSUE TITLE
Multiplicity argument of constructors appear in type errors

### ISSUE THREAD

**Issue by aspiwack** [tweag/ghc#243]
I don't know if it's systematic or only certain type errors. I spotted it in the test `T12785b`, where we see things like:

```
        bound by a pattern with constructor:
                   Branch :: forall a (n2 :: Peano).
                             a -->.(n) HTree n2 (HTree ('S n2) a) -->.(n) HTree ('S n2) a,
```

In that particular case, it's particularly wrong because it is `Branch`-in-a-pattern, which is _not_ multiplicity polymorphic!

It's related to, but probably different than, #239 .

**Comment by monoidal** [comment]
After #254, the situation improves a bit but the issue is not yet fixed.

We're now getting
```
                   Branch :: forall a (n2 :: Peano).
                             a -> HTree n2 (HTree ('S n2) a) -> HTree ('S n2) a,
```
`n` is renamed to `n2` because it collides with the hidden multiplicity variable.

If we enable printing linearity variables, the output is now
```
                   Branch :: forall (n1 :: GHC.Types.Multiplicity) a (n2 :: Peano).
                             a -->.(n1) HTree n2 (HTree ('S n2) a) -->.(n1) HTree ('S n2) a,
```

so at least `n1` is now bound.

**Comment by aspiwack** [comment]
I've mentioned it in the wiki, but I'm starting to think that changing the type in `dataConUserTy` may have been the wrong move (I suspect that it's the root cause of this: when printing a constructor, you print its `dataConUserTy`). Maybe we should split `dataConUserTy` and a new `dataConWrapperTy`, revert `dataConUserTy` and make `dataConWrapperTy` be the one with extended multiplicities. Then be careful of using the right one in each place.

I'm not sure that it's actually the issue, though: it's just a thought at this point.

**Comment by monoidal** [comment]
There are two independent issues here.

1. The error message is about a pattern, but it's printing the type of a multiplicity-polymorphic constructor. Fortunately, this happens only with `-XLinearTypes`.

One solution is to have a variant of `dataConUserType` that does not multiply every field by a multiplicity. I would even make this function output `SDoc` instead of a type - looking at the usage of `dataConUserType`, I think it will be used only in this error message. However, if we do this, then nonusers of linear types will see patterns with `->.` inside them, which necessitates another hack :( We could make this function print all fields as nonlinear, but then, pattern matching on `Id x` and on `Unrestricted x` behaves differently with `XLinearTypes`, and I think this should be reflected in the pattern type.

I would leave this until we have proper linearity + pattern synonym story.

2. The user called the variable `n` but multiplicity is adding another `n` variable, and the user is seeing `n2`. This should be fixable by changing `multiplicityTyVarList`. I haven't checked this, but a different possibility is to restructure tidying and printing: if we removed the multiplicity `n` variable (`defaultNonStandardVars`) before tidying, this would not be a problem to nonusers of Linear Haskell.

(idea for later: in a record `{ x, y :: T }` implicit multiplicity variables should be called `nx, ny` or `xn, yn`).

**Comment by monoidal** [comment]
Item 2 is resolved by #332; item 1 is not a backwards compatibility issue and probably best combined with pattern synonyms. Moving out of backwards compat column.

**Comment by monoidal** [comment]
Item 1 is resolved by introduction of `dataConDisplayType` (#406).
